client/lib/signup/step-actions.js

```diff
@@ -52,62 +53,91 @@ function setThemeOnSite( callback, { siteSlug }, { themeSlug } ) {
-			if ( error ) {
-				return callback( error );
+function addDomainItemsToCart( callback, dependencies, { domainItem, googleAppsCartItem, isPurchasingItem, siteUrl, themeSlug, themeItem } ) {
+	wpcom.undocumented().sitesNew( {
+		blog_name: siteUrl,
+		blog_title: siteUrl,
+		options: {
+			theme: dependencies.theme
+		},
+		validate: false,
+		find_available_url: isPurchasingItem
+	}, function( error, response ) {
+		if ( error ) {
+			callback( error );
+
+			return;
+		}
+
+		const siteSlug = response.blog_details.blogname + '.wordpress.com';
+		const isFreeThemePreselected = themeSlug && ! themeItem;
+		const providedDependencies = {
+			siteSlug,
+			domainItem,
+			themeItem
+		};
+		const addToCartAndProceed = () => {
+			let newCartItems = [];
+
+			if ( domainItem ) {
+				newCartItems = [ ...newCartItems, domainItem ];
 			}
 
-			const siteSlug = response.blog_details.blogname + '.wordpress.com';
-			const isFreeThemePreselected = themeSlug && ! themeItem;
-			const providedDependencies = {
-				siteSlug,
-				domainItem,
-				themeItem,
-			};
-			const addToCartAndProceed = () => {
-				let newCartItems = [];
-
-				if ( domainItem ) {
-					newCartItems = [ ...newCartItems, domainItem ];
-				}
-				if ( googleAppsCartItem ) {
-					newCartItems = [ ...newCartItems, googleAppsCartItem ];
-				}
-				if ( themeItem ) {
-					newCartItems = [ ...newCartItems, themeItem ];
-				}
-
-				if ( newCartItems.length ) {
-					SignupCart.addToCart( siteSlug, newCartItems, function( cartError ) {
-						callback( cartError, providedDependencies );
-					} );
-				} else {
-					callback( [], providedDependencies );
-				}
-			};
-
-			if ( ! user.get() && isFreeThemePreselected ) {
-				return setThemeOnSite( addToCartAndProceed, { siteSlug }, { themeSlug } );
+			if ( googleAppsCartItem ) {
+				newCartItems = [ ...newCartItems, googleAppsCartItem ];
 			}
 
-			if ( user.get() && isFreeThemePreselected ) {
-				return fetchSitesAndUser( siteSlug, setThemeOnSite.bind( this, addToCartAndProceed, { siteSlug }, { themeSlug } ) );
-			} else if ( user.get() ) {
-				return fetchSitesAndUser( siteSlug, addToCartAndProceed );
+			if ( themeItem ) {
+				newCartItems = [ ...newCartItems, themeItem ];
 			}
 
+			if ( newCartItems.length ) {
+				SignupCart.addToCart( siteSlug, newCartItems, function( cartError ) {
+					callback( cartError, providedDependencies );
+				} );
+			} else {
+				callback( undefined, providedDependencies );
+			}
+		};
+
+		if ( ! user.get() && isFreeThemePreselected ) {
+			setThemeOnSite( addToCartAndProceed, { siteSlug }, { themeSlug } );
+		} else if ( user.get() && isFreeThemePreselected ) {
+			fetchSitesAndUser( siteSlug, setThemeOnSite.bind( this, addToCartAndProceed, { siteSlug }, { themeSlug } ) );
+		} else if ( user.get() ) {
+			fetchSitesAndUser( siteSlug, addToCartAndProceed );
+		} else {
 			addToCartAndProceed();
-		} );
+		}
+	} );
+}
+
+function addFreeTrialToCart( callback, dependencies, data ) {
```

Comment:
We now have [an action for starting the free trial](https://github.com/Automattic/wp-calypso/blob/master/client/lib/upgrades/actions/free-trials.js#L40), so instead of adding the free trial to the user's cart and redirecting them to checkout, we can start the trial and redirect the user to `/plans/:site` if they didn't select a domain.